File: src/muhurta/muhurtaTable.ts
// Engine-free reader for a muhurta table.
//
// Ships as a separate entry point (`panchang-ts/muhurta`) that imports NO
// astronomy code, so an app can read pre-computed auspicious dates without
// pulling the engine into its client bundle. This is the member the table family
// was missing: festivals, eclipses and Moon phases each had a builder and a
// reader; muhurta had neither, so `findAuspiciousDates` had to run the full
// engine on device for every query.
//
// **No table is bundled.** Muhurta scoring is location- and rule-dependent, so
// there is no universal table to ship. Build your own with `buildMuhurtaTable`
// (from the main `panchang-ts` entry, which does use the engine), cache the
// JSON, and pass it to the accessors below.

import type {
  MuhurtaFile,
  MuhurtaTableDay,
} from './muhurtaTableTypes';

export type {
  MuhurtaTableLanguage,
  MuhurtaFactor,
  MuhurtaTableDay,
  PackedMuhurtaTableDay,
  MuhurtaTableMeta,
  MuhurtaFile,
} from './muhurtaTableTypes';

/**
 * Inclusive Gregorian year range a table covers.
 *
 * Read straight off `_meta`; provided so callers can range-check without
 * reaching into the file shape.
 */
export function readMuhurtaYearRange(
  source: MuhurtaFile,
): { start: number; end: number } {
  return { start: source._meta.startYear, end: source._meta.endYear };
}

/** The occasion a table was built for. */
export function readMuhurtaOccasion(source: MuhurtaFile): string {
  return source._meta.occasion;
}

function flatten(source: MuhurtaFile, yearKey: string): MuhurtaTableDay[] | null {
  const days = source.years[yearKey];
  if (!days) return null;
  const dict = source._dict;
  return days.map(d => ({
    date: d.date,
    score: d.s,
    passes: d.p === 1,
    // An index outside the dictionary means a corrupt or hand-edited table;
    // drop the factor rather than emit an undefined-shaped object.
    factors: d.f.map(i => dict[i]).filter((f): f is NonNullable<typeof f> => f !== undefined),
  }));
}

/**
 * Scored days for a Gregorian year, ordered by date.
 *
 * Returns `null` if `year` is outside the table's range. If the table was built
 * with `includeFailures: false` (the default), only days that pass the rule are
 * present.
 *
 * @param source Table to read from — a {@link buildMuhurtaTable} result you
 *               built and cached. Required: nothing is bundled.
 * @param year   Gregorian year.
 */
export function readMuhurtaForYear(
  source: MuhurtaFile,
  year: number,
): MuhurtaTableDay[] | null {
  return flatten(source, String(year));
}

/**
 * The scored day for a specific date, or `null` when the date is outside the
 * table's range or was not stored (a failing day in a passes-only table).
 *
 * @param source Table to read from. Required: nothing is bundled.
 * @param date   Either an ISO `YYYY-MM-DD` string (interpreted in the table's
 *               reference timezone) or a `Date` (its local calendar date in the
 *               table's reference timezone is used).
 */
export function readMuhurtaForDate(
  source: MuhurtaFile,
  date: string | Date,
): MuhurtaTableDay | null {
  const key = typeof date === 'string'
    ? date
    : toDateKey(date, source._meta.timezoneOffsetMinutes);
  const days = flatten(source, key.slice(0, 4));
  if (!days) return null;
  return days.find(d => d.date === key) ?? null;
}

/**
 * The best-scoring days in the table, highest first — the query a muhurta table
 * exists to answer.
 *
 * @param source Table to read from.
 * @param limit  Maximum number of days to return. Defaults to 10.
 */
export function readBestMuhurtaDays(
  source: MuhurtaFile,
  limit: number = 10,
): MuhurtaTableDay[] {
  const all: MuhurtaTableDay[] = [];
  for (const yearKey of Object.keys(source.years)) {
    const days = flatten(source, yearKey);
    if (days) all.push(...days);
  }
  // Ties broken by date so the result is deterministic rather than dependent on
  // object key order.
  all.sort((a, b) => (b.score - a.score) || a.date.localeCompare(b.date));
  return all.slice(0, Math.max(0, limit));
}

function toDateKey(d: Date, offsetMinutes: number): string {
  const shifted = new Date(d.getTime() + offsetMinutes * 60_000);
  const y = shifted.getUTCFullYear();
  const m = String(shifted.getUTCMonth() + 1).padStart(2, '0');
  const day = String(shifted.getUTCDate()).padStart(2, '0');
  return `${y}-${m}-${day}`;
}

File: source/go/internal/muhurta/muhurtatable.go
package muhurta

import (
	"sort"
	"strconv"
	"strings"

	"github.com/ishankgupta95/panchang-ts/source/go/v5/internal/types"
)

type MuhurtaYearRange struct {
	Start int `json:"start"`
	End   int `json:"end"`
}

func ReadMuhurtaYearRange(source MuhurtaFile) MuhurtaYearRange {
	return MuhurtaYearRange{Start: source.Meta.StartYear, End: source.Meta.EndYear}
}

func ReadMuhurtaOccasion(source MuhurtaFile) string { return source.Meta.Occasion }

func flatten(source MuhurtaFile, yearKey string) ([]MuhurtaTableDay, bool) {
	days, ok := source.Years[yearKey]
	if !ok {
		return nil, false
	}
	out := make([]MuhurtaTableDay, 0, len(days))
	for _, d := range days {
		factors := make([]MuhurtaFactor, 0, len(d.F))
		for _, i := range d.F {
			if i >= 0 && i < len(source.Dict) {
				factors = append(factors, source.Dict[i])
			}
		}
		out = append(out, MuhurtaTableDay{
			Date: d.Date, Score: d.S, Passes: d.P == 1, Factors: factors,
		})
	}
	return out, true
}

func ReadMuhurtaForYear(source MuhurtaFile, year int) ([]MuhurtaTableDay, bool) {
	return flatten(source, strconv.Itoa(year))
}

func ReadMuhurtaForDateKey(source MuhurtaFile, key string) (MuhurtaTableDay, bool) {
	if len(key) < 4 {
		return MuhurtaTableDay{}, false
	}
	days, ok := flatten(source, key[:4])
	if !ok {
		return MuhurtaTableDay{}, false
	}
	for _, d := range days {
		if d.Date == key {
			return d, true
		}
	}
	return MuhurtaTableDay{}, false
}

func ReadMuhurtaForDate(source MuhurtaFile, dateMs int64) (MuhurtaTableDay, bool) {
	return ReadMuhurtaForDateKey(source, toDateKey(dateMs, source.Meta.TimezoneOffsetMinutes))
}

func ReadBestMuhurtaDays(source MuhurtaFile, limit int) []MuhurtaTableDay {
	all := []MuhurtaTableDay{}
	keys := make([]string, 0, len(source.Years))
	for k := range source.Years {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	for _, yearKey := range keys {
		if days, ok := flatten(source, yearKey); ok {
			all = append(all, days...)
		}
	}
	sort.SliceStable(all, func(i, j int) bool {
		if all[i].Score != all[j].Score {
			return all[i].Score > all[j].Score
		}
		return strings.Compare(all[i].Date, all[j].Date) < 0
	})
	if limit < 0 {
		limit = 0
	}
	if limit > len(all) {
		limit = len(all)
	}
	return all[:limit]
}

func toDateKey(ms int64, offsetMinutes int) string {
	shifted := types.Date(ms + int64(offsetMinutes)*60_000)
	var b strings.Builder
	b.Grow(10)
	b.WriteString(strconv.Itoa(shifted.UTCFullYear()))
	b.WriteByte('-')
	b.WriteString(pad2(shifted.UTCMonth() + 1))
	b.WriteByte('-')
	b.WriteString(pad2(shifted.UTCDate()))
	return b.String()
}

func pad2(n int) string {
	s := strconv.Itoa(n)
	if len(s) < 2 {
		return "0" + s
	}
	return s
}
